jdtls.ext/com.microsoft.jdtls.ext.core/src/org/eclipse/jdt/internal/jarpackager/JarPackageUtil.java

```diff
@@ -0,0 +1,356 @@
+	 * @param   isCompressed				whether the jar is compressed or not
+	 * @param   fJarOutputStream            the destination JarOutputStream
+	 * @param	fDirectories				the temporary set saves existing directories
+	 *
+	 * @throws  IOException 				if an I/O error has occurred
+	 *
+	 * @since 1.14
+	 *
+	 */
+	private static void addFile(File file, IPath path, boolean areDirectoryEntriesIncluded,
+            boolean isCompressed, JarOutputStream fJarOutputStream, Set<String> fDirectories) throws IOException {
+		if (areDirectoryEntriesIncluded)
+			addDirectories(path, fJarOutputStream, fDirectories);
+
+		JarEntry newEntry= new JarEntry(path.toString().replace(File.separatorChar, '/'));
+
+		if (isCompressed)
+			newEntry.setMethod(ZipEntry.DEFLATED);
+			// Entry is filled automatically.
+		else {
+			newEntry.setMethod(ZipEntry.STORED);
+			calculateCrcAndSize(newEntry, new FileInputStream(file), new byte[4096]);
+		}
+
+		newEntry.setTime(file.lastModified());
+		addEntry(newEntry, new FileInputStream(file), fJarOutputStream);
+	}
+
+	/**
+	 * Creates the directory entries for the given path and writes it to the current archive.
+	 * Extracted from org.eclipse.jdt.ui.jarpackager.JarWriter3
+	 *
+	 * @param   destinationPath 		the path to add
+	 * @param   fJarOutputStream        the destination JarOutputStream
+	 * @param	fDirectories			the temporary set saves existing directories
+	 *
+	 * @throws IOException if an I/O error has occurred
+	 *
+	 * @since 1.14
+	 */
+	public static void addDirectories(IPath destinationPath, JarOutputStream fJarOutputStream, Set<String> fDirectories) throws IOException {
+		addDirectories(destinationPath.toString(), fJarOutputStream, fDirectories);
+	}
+
+	/**
+	 * Calculates the crc and size of the resource and updates the entry.
+	 * Extracted from org.eclipse.jdt.internal.ui.jarpackager.JarPackagerUtil
+	 *
+	 * @param entry			the jar entry to update
+	 * @param stream		the input stream
+	 * @param buffer		a shared buffer to store temporary data
+	 *
+	 * @throws IOException	if an input/output error occurs
+	 *
+	 * @since 1.14
+	 */
+	public static void calculateCrcAndSize(final ZipEntry entry, final InputStream stream, final byte[] buffer) throws IOException {
+		int size= 0;
+		final CRC32 crc= new CRC32();
+		int count;
+		try {
+			while ((count= stream.read(buffer, 0, buffer.length)) != -1) {
+				crc.update(buffer, 0, count);
+				size+= count;
+			}
+		} finally {
+			if (stream != null) {
+				try {
+					stream.close();
+				} catch (IOException exception) {
+					// Do nothing
+				}
+			}
+		}
+		entry.setSize(size);
+		entry.setCrc(crc.getValue());
+	}
+
+	/**
+	 * add a warning message into the MultiStatus.
+	 *
+	 * @param message		the message to add
+	 * @param error			the reason of the message
+	 * @param fStatus		the MultiStatus to write
+	 *
+	 * @since 1.14
+	 */
+	protected final static void addWarning(String message, Throwable error, MultiStatus fStatus) {
```

Comment:
Why use protected modifier for static method?

Reply:
The original implementation of `addWarning` is a protected one, I just change it to static. A private should be used here.